cmd/algokey/keyreg.go

```diff
@@ -0,0 +1,264 @@
+	}
+}
+
+func mustConvertB64ToDigest(b64 string) (digest crypto.Digest) {
+	data, err := base64.StdEncoding.DecodeString(b64)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to decode digest '%s': %s\n\n", b64, err)
+		os.Exit(1)
+	}
+	if len(data) != len(digest[:]) {
+		fmt.Fprintf(os.Stderr, "Unexpected decoded digest length decoding '%s'.\n\n", b64)
+		os.Exit(1)
+	}
+	copy(digest[:], data)
+	return
+}
+
+func getGenesisInformation(network string) (string, crypto.Digest, error) {
+	// For testing purposes, there is a secret option to override the genesis information.
+	idOverride := os.Getenv("ALGOKEY_GENESIS_ID")
+	hashOverride := os.Getenv("ALGOKEY_GENESIS_HASH")
+	if idOverride != "" && hashOverride != "" {
+		return idOverride, mustConvertB64ToDigest(hashOverride), nil
+	}
+
+	// Otherwise check that network matches one of the known networks.
+	gen, ok := validNetworks[strings.ToLower(network)]
+	if !ok {
+		return "", crypto.Digest{}, fmt.Errorf("unknown network '%s' provided. Supported networks: %s",
+			network,
+			strings.Join(validNetworkList, ", "))
+	}
+
+	return gen.id, gen.hash, nil
+}
+
+func run(params keyregCmdParams) error {
+	// Implicit last valid
+	if params.lastValid == 0 {
+		params.lastValid = params.firstValid + txnLife
+	}
+
+	if !params.offline {
+		if params.partkeyFile == "" {
+			return errors.New("must provide --partkey-file when registering participation keys")
+		}
+		if params.addr != "" {
+			return errors.New("do not provide --address when registering participation keys")
+		}
+	} else {
+		if params.addr == "" {
+			return errors.New("must provide --address when bringing an account offline")
+		}
+		if params.partkeyFile != "" {
+			return errors.New("do not provide --partkey-file when bringing an account offline")
+		}
+	}
+
+	var accountAddress basics.Address
+	if params.addr != "" {
+		var err error
+		accountAddress, err = basics.UnmarshalChecksumAddress(params.addr)
+		if err != nil {
+			return fmt.Errorf("unable to parse --address: %w", err)
+		}
+	}
+
+	if params.partkeyFile != "" && !util.FileExists(params.partkeyFile) {
+		return fmt.Errorf("cannot access partkey-file '%s'", params.partkeyFile)
+	}
+
+	if util.FileExists(params.txFile) || params.txFile == stdoutFilenameValue {
+		return fmt.Errorf("tx-file '%s' already exists", params.partkeyFile)
+	}
+
+	// Lookup information from partkey file
+	var part *account.Participation
+	if params.partkeyFile != "" {
+		partDB, err := db.MakeErasableAccessor(params.partkeyFile)
+		if err != nil {
+			return fmt.Errorf("cannot open partkey %s: %v", params.partkeyFile, err)
+		}
+
+		partkey, err := account.RestoreParticipation(partDB)
+		if err != nil {
+			return fmt.Errorf("cannot load partkey %s: %v", params.partkeyFile, err)
+		}
+
+		part = &partkey.Participation
+
+		if params.firstValid < uint64(part.FirstValid) {
+			return fmt.Errorf("first-valid (%d) is earlier than the key first valid (%d)", params.firstValid, part.FirstValid)
```

Comment:
could you please rephrase the error message so that it would be clear what is the expectation and what was violated ?

Reply:
( same issue with the rest of the error messages below )

Reply:
I'll reword this one, but the others look good to me